### PR TITLE
Fix Cortex-M registers in bare metal debugging

### PR DESCRIPTION
**Pull request by ttimasdf**
Fix the second issue mentioned in #264, regarding the xSPR register and the thumb mode bit offset.

### CODE REVIEW

File: pwndbg/commands/__init__.py [hunk 1]
```diff
@@ -91,7 +91,11 @@ def check_repeated(self, argument, from_tty):
 
         last_line = lines[-1]
         number, command = last_line.split(None, 1)
-        number = int(number)
+        try:
```
Comment (disconnect3d) [inline pwndbg/commands/__init__.py:94]: Is this a gdb 8.2 hack? Is this needed here?

File: pwndbg/disasm/__init__.py [hunk 2]
```diff
@@ -83,8 +83,16 @@ def get_disassembler(pc):
     extra = None
 
     if pwndbg.arch.current in ('arm', 'aarch64'):
-        extra = {0:CS_MODE_ARM,
-                 0x20:CS_MODE_THUMB}[pwndbg.regs.cpsr & 0x20]
+        if pwndbg.regs.xpsr:
```
Comment (disconnect3d) [inline pwndbg/disasm/__init__.py:86]: Is this condition correct? If `xpsr` would have a value of `0` we will assume the `cpsr` shall be used.

This should probably be `if pwndbg.regs.xpsr is not None:`?

File: pwndbg/disasm/__init__.py [hunk 3]
```diff
@@ -83,8 +83,16 @@ def get_disassembler(pc):
...
+            extra = {
+                0: CS_MODE_MCLASS,
+                0x1000000: CS_MODE_MCLASS | CS_MODE_THUMB,
```
Comment (disconnect3d) [inline pwndbg/disasm/__init__.py:89]: https://github.com/aquynh/capstone/blob/d70e2b286a7466ab8ed6895213acc24e52b8b2fa/bindings/powershell/Capstone/Capstone.psm1#L120 nice

File: pwndbg/disasm/arm.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def condition(self, instruction):
         if instruction.address != pwndbg.regs.pc:
             return False
 
-        cpsr = pwndbg.regs.cpsr
+        cpsr = pwndbg.regs.cpsr or pwndbg.regs.xpsr
```
Comment (disconnect3d) [inline pwndbg/disasm/arm.py:49]: if `cpsr` is 0 we will use xpsr, is it correct behavior?

File: pwndbg/emu/emulator.py [hunk 5]
```diff
@@ -200,7 +200,16 @@ def get_uc_mode(self):
         mode = 0
 
         if arch in ('arm', 'aarch64'):
-            mode |= {0:U.UC_MODE_ARM,0x20:U.UC_MODE_THUMB}[pwndbg.regs.cpsr & 0x20]
+            if pwndbg.regs.xpsr:
```
Comment (disconnect3d) [inline pwndbg/emu/emulator.py:203]: same issue as before: if it is 0 we will go to else

File: pwndbg/commands/context.py [hunk 6]
```diff
@@ -131,6 +131,8 @@ def get_regs(*regs):
             continue
 
         value = pwndbg.regs[reg]
+        if not value and reg.lower() == "cpsr":
+            value = pwndbg.regs["xpsr"]
```
Comment (disconnect3d) [inline pwndbg/commands/context.py:135]: Is this needed here? I don't like that we hardcode a specific register here; can this code be moved elsewhere?

**Review by disconnect3d — CHANGES_REQUESTED** [review]
Hey,

Thanks for the PR! Please see comments.

Also, is there any way to test this out if I don't have a bare metal Cortex-M?
Like some example that we could launch on Qemu?

File: pwndbg/commands/__init__.py (hunk 1 above)
Comment (ttimasdf) [reply pwndbg/commands/__init__.py:94]: Actually when adding this I was hacking around changing this or that wondering what make things work 😅 And it Just Works™
I'll figure out later whether these lines are essential, my device not at hand for the moment.

File: pwndbg/disasm/__init__.py (hunk 2 above)
Comment (ttimasdf) [reply pwndbg/disasm/__init__.py:86]: You are right, value of 0 IS a valid cpsr value and should be considered.

File: pwndbg/disasm/__init__.py (hunk 3 above)
Comment (ttimasdf) [reply pwndbg/disasm/__init__.py:89]: As for this, I'm not sure the flags should be `CS_MODE_MCLASS | CS_MODE_THUMB` or just `CS_MODE_THUMB`. Not yet look into their codes.

File: pwndbg/disasm/arm.py (hunk 4 above)
Comment (ttimasdf) [reply pwndbg/disasm/arm.py:49]: err, right. same as above.

**Comment by disconnect3d** [comment]
@ttimasdf can you update this?

**Comment by ttimasdf** [comment]
Hi, sorry for untimely responds. I've been busy at work recently so didn't keep update. I'll get start when I have time or maybe weekend :cat2:

**Comment by disconnect3d** [comment]
@ttimasdf bump; any chance to start working on this at some point?

**Comment by disconnect3d** [comment]
@ttimasdf bump

**Comment by disconnect3d** [comment]
@ttimasdf ping, any chance to finalize this?:<

**Pull request closed without merging**